From 763296d6423cad091b33291cee8e34c2aea3d0ca Mon Sep 17 00:00:00 2001
From: Antonio Alvarez Feijoo <antonio.feijoo@suse.com>
Date: Thu, 23 Apr 2026 10:18:59 +0200
Subject: [PATCH 8/8] fix(systemd-cryptsetup): load libcryptsetup via dlopen

The `systemd-cryptsetup` binary no longer depends on `libcryptsetup`, and there
is no other binary installed with the `systemd-cryptsetup` or `crypt` modules
that installs it.

```
switch_root:/root# systemd-cryptsetup --version | head -1
systemd 261 (261~devel+git20260423.43dab5ea8)
switch_root:/root# systemd-cryptsetup help
Shared library 'libcryptsetup.so.12' is not available: libcryptsetup.so.12: cannot open shared object file: No such file or directory
switch_root:/root# echo $?
1
```

Required since https://github.com/systemd/systemd/commit/43dab5ea8797e45e0702f8ee89cdf25e577a652b

(cherry picked from commit 60a5e034604d9740d460881ff1d03794b9a9b08a)

Resolves: rhbz#2481748
---
 modules.d/11systemd-cryptsetup/module-setup.sh | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/modules.d/11systemd-cryptsetup/module-setup.sh b/modules.d/11systemd-cryptsetup/module-setup.sh
index 6bb0a42f..09c2082f 100755
--- a/modules.d/11systemd-cryptsetup/module-setup.sh
+++ b/modules.d/11systemd-cryptsetup/module-setup.sh
@@ -42,6 +42,11 @@ depends() {
     return 0
 }
 
+# called by dracut
+config() {
+    add_dlopen_features+=" libsystemd-shared-*.so:cryptsetup "
+}
+
 # called by dracut
 install() {
     # the cryptsetup targets are already pulled in by 00systemd, but not
@@ -56,6 +61,12 @@ install() {
         "$systemdsystemunitdir"/remote-cryptsetup.target \
         "$systemdsystemunitdir"/initrd-root-device.target.wants/remote-cryptsetup.target
 
+    # Install required libraries.
+    if [[ ! $USE_SYSTEMD_DLOPEN_DEPS ]]; then
+        inst_libdir_file \
+            {"tls/$DRACUT_ARCH/",tls/,"$DRACUT_ARCH/",}"libcryptsetup.so.*"
+    fi
+
     if [[ $hostonly ]] && [[ -f $initdir/etc/crypttab ]]; then
         # for each entry in /etc/crypttab check if the key file is backed by a socket unit and if so,
         # include it along with its corresponding service unit.
-- 
2.54.0

